ipld/resolver/src/service.rs

```diff
@@ -1,17 +1,431 @@
+    where
+        S: BitswapStore<Params = P>,
+    {
+        let peer_id = config.network.local_peer_id();
+        let transport = build_transport(config.network.local_key.clone());
+        let behaviour = Behaviour::new(config.network, config.discovery, config.membership, store)?;
+
+        // NOTE: Hardcoded values from Forest. Will leave them as is until we know we need to change.
+
+        let limits = ConnectionLimits::default()
+            .with_max_pending_incoming(Some(10))
+            .with_max_pending_outgoing(Some(30))
+            .with_max_established_incoming(Some(config.connection.max_incoming))
+            .with_max_established_outgoing(None) // Allow bitswap to connect to subnets we did not anticipate when we started.
+            .with_max_established_per_peer(Some(5));
+
+        let swarm = SwarmBuilder::with_tokio_executor(transport, behaviour, peer_id)
+            .connection_limits(limits)
+            .notify_handler_buffer_size(std::num::NonZeroUsize::new(20).expect("Not zero"))
+            .connection_event_buffer_size(64)
+            .build();
+
+        let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
+
+        let service = Self {
+            listen_addr: config.connection.listen_addr,
+            swarm,
+            queries: Default::default(),
+            request_rx: rx,
+            background_lookup_filter: BloomFilter::with_rate(
+                0.1,
+                config.connection.expected_peer_count,
+            ),
+        };
+
+        let client = Client { request_tx: tx };
+
+        Ok((service, client))
+    }
+
     /// Start the swarm listening for incoming connections and drive the events forward.
-    pub async fn run(self) -> anyhow::Result<()> {
-        todo!("IPC-37")
+    pub async fn run(mut self) -> anyhow::Result<()> {
+        // Start the swarm.
+        Swarm::listen_on(&mut self.swarm, self.listen_addr.clone())?;
+
+        loop {
+            select! {
+                swarm_event = self.swarm.next() => match swarm_event {
+                    // Events raised by our behaviours.
+                    Some(SwarmEvent::Behaviour(event)) => {
+                        self.handle_behaviour_event(event)
+                    },
+                    // Connection events are handled by the behaviours, passed directly from the Swarm.
+                    Some(_) => { },
+                    // The connection is closed.
+                    None => { break; },
+                },
+                request = self.request_rx.recv() => match request {
+                    // A Client sent us a request.
+                    Some(req) => self.handle_request(req),
+                    // All Client instances have been dropped.
+                    // We could keep the Swarm alive to serve content to others,
+                    // but we ourselves are unable to send requests. Let's treat
+                    // this as time to quit.
+                    None => { break; }
```

Comment:
In order to spawn a client, the `Service` would have to have a copy of the `Sender<Request>` channel (which is what happens in Forest. But the `Service` instance is consumed by `run`, so you can't call anything on it any more, you can't have a reference to it, you would not be able to clone a new `Sender`. And by itself having a copy of `Sender`, it would never experience this condition either, unless it manually drop it. 

So I decided that a cleaner way is for `Service::new` to return the `Service` and the `Sender` (wrapped into a `Client`). Then you can make as many clones of the `Client` as you want, before or after you `run` the `Service`. 

But you are right, you have to make sure not to drop all `Client` instances. But this is easy, because if you did, and you needed another one, you wouldn't be able to get it from anywhere, so the compiler wouldn't let this happen. And if you never ever need to make a `Client` call (after telling it what subnets you support), then yeah, you just need to make sure you keep an instance in your application state anyway.